File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) 
    {
        if(p == nullptr && q != nullptr || p != nullptr && q == nullptr) return false;
        else if(p == nullptr && q == nullptr) return true;
        return p->val == q->val && isSameTree(p->left, q->left) && isSameTree(p->right, q->right);
    }
    
    bool isSubtree(TreeNode* root, TreeNode* subRoot) 
    {
        if(isSameTree(root, subRoot)) return true;
        
        return dfs(root, subRoot);
    }
    
    bool dfs(TreeNode* root, TreeNode* subRoot)
    {
        if(root == nullptr && subRoot != nullptr || root != nullptr && subRoot == nullptr) return false;
        
        if(isSameTree(root, subRoot)) return true;
        
        return dfs(root->left, subRoot) || dfs(root->right, subRoot);
    }
};

File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sameTree(self, A, B):
        if A and not B or not A and B:
            return False
        if not A and not B:
            return True
        
        return A.val==B.val and self.sameTree(A.left, B.left) and self.sameTree(A.right, B .right)
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not root and subRoot or root and not subRoot:
            return False
        return self.sameTree(root, subRoot) or self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)
        